### ISSUE TITLE
NPE in Semantic Analyzer `Types.getReferredType()`

### ISSUE THREAD

**Issue by IMS94** [ballerina-platform/ballerina-lang#35868]
**Description:**
When we copy and paste the following code into an empty bal file, we get an NPE.
```ballerina
public function main() returns error?{
   
   http:Client nobelPrizes = check new(API_URL);
   LaureateResults laureateResults = check nobelPrizes->get("/laureate.json");
   Laureate[] laureates = laureateResults.laureates;
 
   //Query expression to destructure the Laureate records to the preferred format
   Laureate[] formattedLaureates = 
      from Laureate laureate in laureates
      let boolean isAlive = (laureate.died == "0000-00-00")
      select {id:laureate.id,firstname:laureate.firstname,
surname:laureate.surname?:"",
             bornCountry:laureate.bornCountry?:"", gender:laureate.gender?:"", 
             prizes:laureate.prizes, "isAlive":isAlive};
 
   io:print(formattedLaureates);
}
```

Error log:
```java
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in
https://github.com/ballerina-platform/ballerina-lang/issues with the
log you get below and your sample code.

We thank you for helping make us better.

[2022-04-21 01:31:43,784] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.getReferredType(Types.java:1310)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types$BSameTypeVisitor.visit(Types.java:2501)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types$BSameTypeVisitor.visit(Types.java:2490)
	at org.wso2.ballerinalang.compiler.semantics.model.types.BType.accept(BType.java:86)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isSameType(Types.java:352)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isSameType(Types.java:317)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isAssignable(Types.java:674)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isAssignable(Types.java:662)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.handleInputClauseVariables(TypeChecker.java:5635)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:5502)
	at org.wso2.ballerinalang.compiler.tree.clauses.BLangFromClause.accept(BLangFromClause.java:43)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.lambda$visit$26(TypeChecker.java:5257)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:5257)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangQueryExpr.accept(BLangQueryExpr.java:105)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:380)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:354)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1052)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:54)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4060)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4031)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4023)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1907)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:50)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4060)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4031)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4027)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:524)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:60)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4060)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:500)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:76)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4060)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4031)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4023)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:349)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:309)
	at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:188)
	at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:114)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:436)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:386)
	at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:213)
	at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:198)
	at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:114)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:103)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:214)
	at io.ballerina.projects.Package.getCompilation(Package.java:141)
	at org.ballerinalang.langserver.workspace.BallerinaWorkspaceManager.waitAndGetPackageCompilation(BallerinaWorkspaceManager.java:271)
	at org.ballerinalang.langserver.util.LSClientUtil.chekAndRegisterCommands(LSClientUtil.java:78)
	at org.ballerinalang.langserver.BallerinaTextDocumentService.didChange(BallerinaTextDocumentService.java:556)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Steps to reproduce:**
See description

**Affected Versions:**
2201.0.3 RC2

**Comment by pcnfernando** [comment]
This seems to have been fixed on 2201.3.0 rc1. @IMS94 can you please check

**Comment by KavinduZoysa** [comment]
This is fixed in 3.0 and above.
